pkg/types/main.go

```diff
@@ -16,6 +18,41 @@ const (
 
 type Uint128 C.tb_uint128_t
 
+func (value Uint128) String() string {
+	bytes := *(*[16]byte)(unsafe.Pointer(&value))
+	s := hex.EncodeToString(bytes[:16])
+
+	// Prettier to drop preceeding zeros so you get "0" instead of "0000000000000000"
+	lastNonZero := 0
+	for s[lastNonZero] == '0' && lastNonZero < len(s)-1 {
+		lastNonZero++
+	}
+	return s[lastNonZero:]
+}
+
+// HexStringToUint128 converts a hex-encoded integer to a Uint128.
+func HexStringToUint128(value string) (Uint128, error) {
```

Comment:
> Would it make sense to pass a [16]byte instead of a string? That sidesteps any confusion around lengths — e.g.

Sure can do. I went back and forth between usability and correctness but idk what the right answer is.

> Likewise instead of String(), have Bytes() [16]byte 

String() is a go convention that calls to fmt.Printf look for. We could also have a Bytes() though if you want

Reply:
Lots of example code is going to have this 

```
// Pad string to two hex digits
if len(value)%2 == 1 {
  value = "0" + value
}
```

though if we do the [16]byte input rather than `string` since constructing literal `[16]byte` is not as convenient as constructing literal strings.

My personal preference would be more that we panic if the length is not 16.

Reply:
Maybe I'll add both HexStringToUint128 and HexBytesToUint128.

Reply:
> Sure can do. I went back and forth between usability and correctness but idk what the right answer is.

Our goal for these client implementations (at least for now) is that they are as low-level and unopinionated as possible, so prioritize correctness over convenience. (Imagine that a (more opinionated) higher level client could build on top of it for convenience.)

For example code, `[16]byte` can be generated via Golang's RNG, which is more realistic anyway (i.e. it is what most applications will do).

Reply:
The issue with representing a UUID as a string is there are many different conventions / formats for doing so. We can leave that decision to the application and just use `[16]byte`, which is unambigous.

Reply:
Yup I agree with you. Now I'm just talking about keeping a helper HexStringToUint128 function.

This is also what King already had: https://github.com/tigerbeetledb/tigerbeetle-go/blob/main/tb_client_test.go#L23-L28.

Reply:
Ooh, sorry, I see. Yeah, that is fine, and I agree that we should panic for `len ≠ 16`.

Reply:
Actually there are two things I realize now. There's getting from a true [16]byte to a Uint128 which is just a simple unsafe cast.

There's also how to get from a hex string to the uint128 which involves the hex decoding. New helper coming in.

Reply:
@sentientwaffle does it now make sense? (with all the new functions)

Reply:
Yep, looks good.